---
title: "griph: Graph Inference of Population Heterogeneity"
author: "Panagiotis Papasaikas, Michael Stadler"
date: "`r BiocStyle::doc_date()`"
package: "`r BiocStyle::pkg_ver('griph')`"
abstract: >
    griph identifies cell types in single cell RNA-seq datasets,
    allowing to control for unwanted sources of variance (e.g. cell cycle)
    that may confound the cell types.

    The most important function that combines all the steps of cell type identification
    from griph is \code{\link{SC_cluster}}. The vignette illustrates its use
    applied to several publicly available datasets.
output: 
    BiocStyle::html_document:
        keep_md: yes
    github_document: default
vignette: >
    %\VignetteIndexEntry{griph: Graph Inference of Population Heterogeneity}
    %\VignetteEngine{knitr::rmarkdown}
    %\VignetteEncoding{UTF-8}
    \usepackage[utf8]{inputenc}
---

# Introduction
Graph Inference of Population Heterogeneity


# Quickstart: A sample analysis
We will use the mouse ES cell data from Buettner et al. for this example (see details
in the "Sample datasets" section). The dataset is included in the package, including
know cell cycle labels:  
```{r buettner_data}
M <- readRDS(system.file("extdata", "buettner_top10k.rds", package = "griph"))
dim(M) # genes by cells
trueLabel <- attr(M, "label")
table(trueLabel)
```

Cell types can be identified using \code{\link{SC_cluster}}:  
```{r buettner_types}
library(griph)
res <- SC_cluster(M, ClassAssignment = trueLabel, plotG = TRUE, fsuffix='buettner')
table(res$MEMB)
```
```{r movegraph, echo=FALSE, results='hide'}
  files <- list.files("./")
  index <- grep(".png", files)
  file.rename(file.path("./", files[index]),
              file.path("./griph_files/figure-html", files[index]))
```


The automatic generation of a graph plot (\code{plotG} and \code{plotSP} arguments) is
switched on here (plotG=TRUE). When plotG is set to TRUE plotGRAO contais an optimized for rendering version of the complete graph and a pdf file of this graph plot is automatically generated:

![Buettner Graph](https://raw.githubusercontent.com/ppapasaikas/griph/master/griph/vignettes/griph_files/figure-html/graph_buettner.png)


In plotGRAO weak edges are pruned, vertex attributes are added for visualization of predicted and known class assignments (if given) and a subset of the vertices is sampled if the graph exceeds the maxG argument. When plotG is set to FALSE plotGRAO returns NULL. In both cases the complete graph object (though missing all plotting-related vertex attributes) is returned in the GRAO slot of the results (e.g here in res$GRAO).


Alternatively (and preferably for large - typically >1000 - numbers of cells ) results can be visualized by applying a dimensionality reduction/projection technique such as tSNE to the affinity
matrix returned by \code{griph}:  
```{r buettner_tsne, out.height="600px", out.width="600px", fig.show='hide'}
library(Rtsne)
out <- Rtsne(as.dist(1-res$DISTM), pca = FALSE, perplexity = 5, is_distance = TRUE)
plot(out$Y, xlab="tSNE-1", ylab="tSNE-2", col=res$MEMB, pch=as.numeric(trueLabel))
```

![Buettner tSNE](https://raw.githubusercontent.com/ppapasaikas/griph/master/griph/vignettes/griph_files/figure-html/buettner_tsne-1.png)

\code{\link{SC_cluster}} identified `r length(unique(res$MEMB))`
cell types in the data, and the confusion matrix returned in the result summarizes
how they relate to the known cell types or states:  
```{r buettner_eval}
res$ConfMatrix # confusion matrix
res$miscl # misclassification error
```

When comparing different classifications with each other, they may differ in granularity,
yet be consistent with each other. This can be seen for example in the above confusion
matrix for predicted classes *3*, *4* and *5* that all contain mostly *G1* cells.
\code{griph} takes this into account when calculating classification error by assigning
each identified class to the major known class (*G1* in this example) and only counting
the 6 cells in *3*, *4* and *5* as wrongly classified that are not *G1*:  




# Sample datasets included in this package  

## Buettner et al., Nature Biotechnology 2015
Computational analysis of cell-to-cell heterogeneity in single-cell RNA-sequencing data reveals hidden subpopulations of cells. Buettner F, Natarajan KN, Casale FP, Proserpio V, Scialdone A, Theis FJ, Teichmann SA, Marioni JC, Stegle O. Nat Biotechnol. 2015 Feb;33(2):155-60. doi: 10.1038/nbt.3102.  

- Description: 288 mouse ES cells (Fluidigm C1), classified into three cell cycle stages
- PubMed: http://www.ncbi.nlm.nih.gov/pubmed/25599176
- DOI: https://doi.org/10.1038/nbt.3102
- Data: http://www.ebi.ac.uk/arrayexpress/experiments/E-MTAB-2805/
- Filtering: all cells, top 10,000 genes

```{r Buettner}
fname <- system.file("extdata", "buettner_top10k.rds", package = "griph")
M <- readRDS(fname)
dim(M)
label <- attr(M, "label")
table(label)
```


## Kolodziejck et al., Cell Stem Cell 2015
Single Cell RNA-Sequencing of Pluripotent States Unlocks Modular Transcriptional Variation. Kolodziejczyk AA, Kim JK, Tsang JC, Ilicic T, Henriksson J, Natarajan KN, Tuck AC, Gao X, Bühler M, Liu P, Marioni JC, Teichmann SA. Cell Stem Cell. 2015 Oct 1;17(4):471-85. doi: 10.1016/j.stem.2015.09.011.

- Description: 704 mouse ES cells (Fluidigm C1), three culture conditions: serum + LIF (lif), 2i + LIF (2i) and alternative 2i + LIF (a2i)
- PubMed: https://www.ncbi.nlm.nih.gov/pubmed/26431182  
- DOI: https://doi.org/10.1016/j.stem.2015.09.011  
- Data: http://www.ebi.ac.uk/arrayexpress/experiments/E-MTAB-2600/, http://www.ebi.ac.uk/teichmann-srv/espresso  

```{r Kolodziejck}
fname <- system.file("extdata", "kolodziejck_top10k.rds", package = "griph")
M <- readRDS(fname)
dim(M)
label <- attr(M, "label")
table(label)
```

## Usoskin et al., Nat Neurosci. 2015
Unbiased classification of sensory neuron types by large-scale single-cell RNA sequencing. Usoskin D, Furlan A, Islam S, Abdo H, Lönnerberg P, Lou D, Hjerling-Leffler J, Haeggström J, Kharchenko O, Kharchenko PV, Linnarsson S, Ernfors P. Nat Neurosci. 2015 Jan;18(1):145-53. doi: 10.1038/nn.3881.

- Description: 799 mouse cells (622 neurons) from lumbar DRGs (custom protocol), different sensory neuron subtypes
- PubMed: https://www.ncbi.nlm.nih.gov/pubmed/25420068  
- DOI: https://doi.org/10.1038/nn.3881  
- Data: https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE59739  

```{r Usoskin}
fname <- system.file("extdata", "usoskin_top10k.rds", package = "griph")
M <- readRDS(fname)
dim(M)
label <- attr(M, "label")
table(label)
label2 <- attr(M, "label2")
table(label2)
label3 <- attr(M, "label3")
table(label3)
```

## Zeisel et al., Science 2015
Brain structure. Cell types in the mouse cortex and hippocampus revealed by single-cell RNA-seq. Zeisel A, Muñoz-Manchado AB, Codeluppi S, Lönnerberg P, La Manno G, Juréus A, Marques S, Munguba H, He L, Betsholtz C, Rolny C, Castelo-Branco G, Hjerling-Leffler J, Linnarsson S. Science. 2015 Mar 6;347(6226):1138-42. doi: 10.1126/science.aaa1934.

- Description: 3005 mouse brain cells (Fluidigm C1), different neuron subtypes
- PubMed: https://www.ncbi.nlm.nih.gov/pubmed/25700174  
- DOI: https://doi.org/10.1126/science.aaa1934  
- Data: http://linnarssonlab.org/cortex/, https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE60361  

```{r Zeisel}
fname <- system.file("extdata", "zeisel_top10k.rds", package = "griph")
M <- readRDS(fname)
dim(M)
label <- attr(M, "label")
table(label)
label2 <- attr(M, "label2")
table(label2)
```



# Session info {.unnumbered}
Here is the output of sessionInfo() on the system on which this document was compiled:
```{r sessionInfo, eval=TRUE, echo=FALSE}
sessionInfo()
```



```{r setup_documentation, echo=FALSE, message=FALSE, results='hide'}
#### Copy the created html to master root as index.html
#### and the md as README.md
  file.copy("griph.html","../../index.html",overwrite=TRUE)
  file.copy("griph.md","../../README.md",overwrite=TRUE)
```

